[ASTImporter] Add importer specific lookup

martong opened the pull request:
There are certain cases when normal C/C++ lookup (localUncachedLookup)
does not find AST nodes. E.g.:
Example 1:
```
  template <class T>
  struct X {
    friend void foo(); // this is never found in the DC of the TU.
  };
```
Example 2:
```
  // The fwd decl to Foo is not found in the lookupPtr of the DC of the
  // translation unit decl.
  struct A { struct Foo *p; };
```
In these cases we create a new node instead of returning with the old one.
To fix it we create a new lookup table which holds every node and we are
not interested in any C++ specific visibility considerations.
Simply, we must know if there is an existing Decl in a given DC.

Fixes #503

balazske commented on include/clang/AST/ASTImporter.h:
```diff
@@ -56,12 +57,19 @@ class TypeSourceInfo;
+    /// Pointer to the import specific lookup table, which may be shared
+    /// amongst several ASTImporter objects.
+    /// If not set then the original C/C++ lookup is used.
+    ASTImporterLookupTable *LookupTable = nullptr;
```
Can we not use a shared pointer for this?

balazske replied:
In the current way of usage shared pointer is not needed. But a notice can be here that the lookup table is an externally managed resource (and should exist during the lifetime of the ASTImporter object). This variable can be an optional too (if it is possible that it contains a reference to ensure non-null value).

balazske commented on include/clang/AST/ASTImporter.h:
```diff
@@ -56,12 +57,19 @@ class TypeSourceInfo;
   /// Imports selected nodes from one AST context into another context,
   /// merging AST nodes where appropriate.
   class ASTImporter {
+    friend class ASTNodeImporter;
```
This friend is not necessary. The `ASTNodeImporter` is already a "private" class and the needed things (for ASTImporter) can be public. (And without friend there is no reference to ASTNodeImporter from any other file.)

martong replied:
AFAIK nested classes are not automatically friends (unfortunately). Also I have made `FindDeclsInToCtx` private in `ASTImporter` so that could be used only inside `ASTNodeImporter`, since that is an implementation detail.

balazske commented on lib/AST/ASTImporterLookupTable.cpp:
```diff
@@ -0,0 +1,103 @@
+
+ASTImporterLookupTable::ASTImporterLookupTable(TranslationUnitDecl &TU) {
+  Builder B(*this);
+  B.TraverseDecl(cast<Decl>(&TU));
```
Is this cast necessary?

martong replied:
Good catch.

balazske commented on lib/AST/ASTImporterLookupTable.cpp:
```diff
@@ -0,0 +1,103 @@
+void ASTImporterLookupTable::remove(NamedDecl *ND) {
+  assert(ND);
+  DeclList &Decls = lookupTable[ND->getDeclContext()][ND->getDeclName()];
+  auto EraseResult = Decls.erase(ND);
```
`bool` is better here than `auto`? (It is compared later with `true`.)

balazske commented on include/clang/CrossTU/CrossTranslationUnit.h:
```diff
@@ -147,6 +148,7 @@ class CrossTranslationUnitContext {
   bool isInvalidFunction(const FunctionDecl *FD);
 
 private:
+  void lazyInitLookupTable(TranslationUnitDecl* ToTU);
```
`*` is at wrong place

balazske replied:
This is still not resolved.

balazske commented on include/clang/AST/ASTImporterLookupTable.h:
```diff
@@ -0,0 +1,47 @@
+  using DeclList = llvm::SmallPtrSet<NamedDecl *, 2>;
+  using NameMap = llvm::SmallDenseMap<DeclarationName, DeclList, 4>;
+  using DCMap = llvm::DenseMap<DeclContext *, NameMap>;
+  DCMap lookupTable;
```
`LookupTable` ?

balazske commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -353,13 +356,24 @@ class ASTImporterTestBase : public CompilerOptionSpecificTest {
 
-  void lazyInitToAST(Language ToLang) {
+  // Initialize the lookup table if not initialized already.
+  void lazyInitLookupTable(TranslationUnitDecl* ToTU) {
```
`*` is at wrong place

Xazax-hun commented on include/clang/AST/ASTImporterLookupTable.h:
```diff
@@ -0,0 +1,47 @@
+class ASTImporterLookupTable {
+  using DeclList = llvm::SmallPtrSet<NamedDecl *, 2>;
+  using NameMap = llvm::SmallDenseMap<DeclarationName, DeclList, 4>;
+  using DCMap = llvm::DenseMap<DeclContext *, NameMap>;
```
Whe do we need a two level map? Could we just have one bigger? I think it might be worth to describe the reason for this choice in a comment.

Xazax-hun commented on include/clang/AST/ASTImporterLookupTable.h:
```diff
@@ -0,0 +1,47 @@
+public:
+  ASTImporterLookupTable(TranslationUnitDecl &TU);
+  void add(NamedDecl *ND);
+  void remove(NamedDecl *ND);
```
Why do we ever need to remove a Decl from the lookup table?

martong replied:
This is only needed when we test. I was pondering to put this as a friend free function in the Test.cpp, but putting it in the class seemed to be more cohesive.

Xazax-hun commented on include/clang/CrossTU/CrossTranslationUnit.h:
```diff
@@ -159,6 +161,7 @@ class CrossTranslationUnitContext {
   llvm::SmallPtrSet<const FunctionDecl *, 8> InvalidFunctions;
   CompilerInstance &CI;
   ASTContext &Context;
+  std::unique_ptr<ASTImporterLookupTable> LookupTable;
```
I think we could have a regular field here (unless we want to create this lazily.)

martong replied:
Well, we have to create that lazily.

Xazax-hun reviewed (commented):
Do we want to include everything in the huge lookup table? I am thinking about static declarations, and declarations in the anonymous namespace. We might not want to find those during our lookup. What do you think?

Xazax-hun commented:
Also, I think it would be great to add some high level documentation somewhere, why do we need a separate kind of lookup.

balazske commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -4547,6 +4656,321 @@ TEST_P(ASTImporterOptionSpecificTestBase,
+  // So we must dig up the underlying CXXRecordDecl.
+  ASTImporterLookupTable LT(*ToTU);
+  auto *FriendD = FirstDeclMatcher<FriendDecl>().match(ToTU, friendDecl());
+  const RecordDecl* RD = getRecordDeclOfFriend(FriendD);
```
`*` is at wrong place

balazske commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -4547,6 +4656,321 @@ TEST_P(ASTImporterOptionSpecificTestBase,
+  auto *FriendD0 = FirstDeclMatcher<FriendDecl>().match(ToTU, friendDecl());
+  auto *FriendD1 = LastDeclMatcher<FriendDecl>().match(ToTU, friendDecl());
+  const RecordDecl* RD0 = getRecordDeclOfFriend(FriendD0);
+  const RecordDecl* RD1 = getRecordDeclOfFriend(FriendD1);
```
`*` at wrong place

balazske commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -4547,6 +4656,321 @@ TEST_P(ASTImporterOptionSpecificTestBase,
+  const RecordDecl* RD0 = getRecordDeclOfFriend(FriendD0);
+  const RecordDecl* RD1 = getRecordDeclOfFriend(FriendD1);
+  ASSERT_EQ(RD0, RD1);
+  ASSERT_EQ(RD1, X);
```
Is `ASSERT_EQ` correct here? (`RD0` and `RD1` are `A` and `B` in the code that should be different, and should be different from X too.)

martong replied:
RD0 and RD1 both refer to the CXXRecordDecl of the target type of the FriendDecl.
Here, `friend struct X` does not create a new forward decl, just references an existing.
See the dump:
```
|-CXXRecordDecl 0xf7edd8 <output.cc:2:7, col:14> col:14 struct X
|-CXXRecordDecl 0xf7ee90 <line:3:7, line:5:7> line:3:14 struct A definition
| |-DefinitionData pass_in_registers empty aggregate standard_layout trivially_copyable pod trivial literal has_constexpr_non_copy_move_ctor can_const_default_init
| | |-DefaultConstructor exists trivial constexpr needs_implicit defaulted_is_constexpr
| | |-CopyConstructor simple trivial has_const_param needs_implicit implicit_has_const_param
| | |-MoveConstructor exists simple trivial needs_implicit
| | |-CopyAssignment trivial has_const_param needs_implicit implicit_has_const_param
| | |-MoveAssignment exists simple trivial needs_implicit
| | `-Destructor simple irrelevant trivial needs_implicit
| |-CXXRecordDecl 0xf7efb8 <col:7, col:14> col:14 implicit struct A
| `-FriendDecl 0xf7f0a8 <line:4:9, col:23> col:16 'struct X':'X'
`-CXXRecordDecl 0xf7f100 <line:6:7, line:8:7> line:6:14 struct B definition
  |-DefinitionData pass_in_registers empty aggregate standard_layout trivially_copyable pod trivial literal has_constexpr_non_copy_move_ctor can_const_default_init
  | |-DefaultConstructor exists trivial constexpr needs_implicit defaulted_is_constexpr
  | |-CopyConstructor simple trivial has_const_param needs_implicit implicit_has_const_param
  | |-MoveConstructor exists simple trivial needs_implicit
  | |-CopyAssignment trivial has_const_param needs_implicit implicit_has_const_param
  | |-MoveAssignment exists simple trivial needs_implicit
  | `-Destructor simple irrelevant trivial needs_implicit
  |-CXXRecordDecl 0xf7f228 <col:7, col:14> col:14 implicit struct B
  `-FriendDecl 0xf7f2e0 <line:7:9, col:23> col:16 'struct X':'X'
```
In contrast to the case of friend functions, where there will be a new forward decl:
```
`-CXXRecordDecl 0xafed88 <output.cc:2:7, col:34> col:13 class Y definition
  |-DefinitionData pass_in_registers empty aggregate standard_layout trivially_copyable pod trivial literal has_constexpr_non_copy_move_ctor can_const_default_init
  | |-DefaultConstructor exists trivial constexpr needs_implicit defaulted_is_constexpr
  | |-CopyConstructor simple trivial has_const_param needs_implicit implicit_has_const_param
  | |-MoveConstructor exists simple trivial needs_implicit
  | |-CopyAssignment trivial has_const_param needs_implicit implicit_has_const_param
  | |-MoveAssignment exists simple trivial needs_implicit
  | `-Destructor simple irrelevant trivial needs_implicit
  |-CXXRecordDecl 0xafeea8 <col:7, col:13> col:13 implicit class Y
  `-FriendDecl 0xaff080 <col:17, col:31> col:29
    `-FunctionDecl 0xafef90 parent 0xac46b8 <col:17, col:31> col:29 F 'void ()'
```

balazske commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -4547,6 +4656,321 @@ TEST_P(ASTImporterOptionSpecificTestBase,
+  DeclarationName Name = RD->getDeclName();
+  auto Res = LT.lookup(ToTU, Name);
+  EXPECT_EQ(Res.size(), 1u);
+  EXPECT_EQ(*Res.begin(), RD);
```
Maybe check that `F` is not found in `Y` by the lookup?

martong commented:
Update: 
We discussed with @balazske that the cases when we compare an in-class method decl with an out-of-class method decl are signs of redecl chain errors.
So I reverted the change related to in-class vs out-of-class check and fixed the redecl chain error rather, please see the commits.

martong commented:
@Xazax-hun 
> Do we want to include everything in the huge lookup table? I am thinking about static declarations, and declarations in the anonymous namespace. We might not want to find those during our lookup. What do you think?

We have to find a previous static declaration from the same TU. Fortunately we already have a solution for that, we do filter the lookup results based on their visibility: See https://github.com/Ericsson/clang/pull/496

Commit:
Add importer specific lookup

Commit:
Set redecl chain of functions before any other import

FunctionDecl import starts with a lookup and then we create a new Decl.
Then in case of CXXConstructorDecl we further import other Decls
(base classes, members through CXXConstructorDecl::inits()) before connecting
the redecl chain.  During those in-between imports structural eq fails
because the canonical decl is different.  This commit fixes this.
Synthesizing a test seemed extremely hard, however, Xerces analysis
reproduces the problem.

Commit:
Import overrides before importing the rest of the chain

During method import we check for structural eq of two methods.
In the structural eq check we check for their isVirtual() flag. That
flag, however, may depend on the number of overrides. Before this
change we imported the overrides *after* we had imported the rest of the
redecl chain.  So, during the import of another decl from the chain
IsVirtual() gave false result.

Writing tests for this is not really possible, because there is no way
to remove an overridden method via the AST API.
(We should access the private ASTContext::OverriddenMethods container.)
And we should do that in the middle of the import function.

martong commented:
Update:
I have updated the PR, so it contains now 3 commits:
1) Add importer specific lookup
2) Set redecl chain of functions before any other import
3) Import overrides before importing the rest of the chain

2 and 3 are required in order to pass the Xerces analysis without assertions. They could go into a separate PR too , but then this PR and those too would not pass the CI.
These are relatively very small changes, I've put them into separate commits to make it easier to upstream them independently.

martong commented:
@balazske 
Ok, changed that and also changed the order of commits, so if there are any more comments about the new lookup i can change them easily.

balazske commented:
I am interested in if the following code is still needed (this is in `VisitRecordDecl` and similar in `VisitFunctionDecl`). The tests show that the lookup result should contain both the templates and the templated decls (is this true in all kind of cases?) therefore this code could be omitted:
```
      if (D->getDescribedTemplate()) {
        if (auto *Template = dyn_cast<ClassTemplateDecl>(Found))
          Found = Template->getTemplatedDecl();
        else
          continue;
      }
```

martong commented:
That is a good catch! So I changed it, but I was a bit worried that it may wreak havoc some lldb tests, because in case of lldb we still use the old C/C++ lookup. However, it looks like the lldb tests do pass, perhpas because they don't really import templates.

balazske reviewed (commented):
I found no other issues, but I am not sure if the last comment about `*` character wrong position was fixed. I can not find the review comment any more but see only "changes requested".

martong commented on include/clang/CrossTU/CrossTranslationUnit.h:
```diff
@@ -147,6 +148,7 @@ class CrossTranslationUnitContext {
   bool isInvalidFunction(const FunctionDecl *FD);
 
 private:
+  void lazyInitLookupTable(TranslationUnitDecl *ToTU);
```
@balazske This was the line with the `*` at wrong place. It is corrected now. It is my bad that I did an amend and then a force push, next time I'll create a separate commit even for the smallest changes to make review easier.

martong commented:
@balazske Thanks for the review. I have squashed the last two commits into one.
